Add Compressor interface

lucagiac81 opened the pull request:
This PR refactors the compression code by introducing a Compressor interface and a CompressorRegistry.
PR #6717 depends on this PR and adds support for custom compressors and dynamic loading.

### Compressor

The Compressor class defines an interface for each compression algorithm to implement. 
A class derived from Compressor is provided for each supported algorithm (SnappyCompressor, LZ4Compressor...).

A compressor has 
- A unique name
- Compress/Uncompress methods
- Optional methods to handle dictionary/context (implemented by zstd, for example)

### CompressorRegistry

The CompressorRegistry holds instances of all available compressors, and it assigns a numeric type to each compressor (following the CompressionType enum).
Compressors for the algorithms supported in RocksDB are added upon registry instantiation.

### Summary of Code Changes

- Add Compressor and CompressorRegistry classes
- Provide a Compressor-derived class for each supported compression algorithm
- Generalize zstd-specific operations in CompressionDict, UncompressionDict, CompressionContext, UncompressionContext
  - Move zstd-specific code to ZSTDCompressor
- Move CompressionDict, UncompressionDict, CompressionContext, UncompressionContext, CompressionInfo, UncompressionInfo to public API
  - The types are used in the Compressor interface, which is public (to eventually support custom compressors in PR #6717)
- Replace switch statements to select compressors with queries to CompressorRegistry

Note about Uncompress function signature in Compressor
- The Uncompress function in Compressor returns data as char* rather than CacheAllocationPtr. A CacheAllocationPtr is created in UncompressData.
- This avoids exposing CacheAllocationPtr in the public API (Compressor has to be public to eventually support custom compressors).
- Allocate and Deallocate functions are added to simplify MemoryAllocator usage by compressors (similar to what is provided with CacheAllocationPtr).

mrambacher commented on include/rocksdb/compressor.h:
```diff
@@ -0,0 +1,206 @@
+  const uint64_t sample_for_compression_;
+
+ public:
+  CompressionInfo(const CompressionOptions& _opts,
+                  const CompressionContext& _context,
+                  const CompressionDict& _dict, CompressionType _type,
+                  uint64_t _sample_for_compression)
```
Rather than taking in a "CompressionOptions" and "CompressionType" should it take in a Compressor?  Why wouldn't we create. Compressor based on these arguments and return it?

Same goes for UncompressionInfo.

lucagiac81 replied:
CompressionOptions and CompressionType carry sufficient information and I didn't change them.

The CompressionType in CompressionInfo/UncompressionInfo is used to get a Compressor from the registry in the CompressData/UncompressData functions (where Compress/Uncompress are eventually called). It is also used in a few other functions, such as CompressBlock and UncompressBlockContents (only the type is needed there). 
A related question is whether we should get the Compressor earlier in the call chain rather than waiting till the last moment. I didn't see a benefit, and it would have triggered more code changes.

The CompressionOptions in CompressionInfo are used to pass options to the Compress method. This part didn't change. Those options are not meant to be stored in the Compressor state. As a Compressor instance is shared across all threads, applying the options to the Compressor would immediately impact any threads using it (e.g., any threads in the middle of a compaction). On the other hand, CompressionOptions are captured at the beginning of each compaction (by GetCompressionOptions) and stay unchanged for the duration of the job. 
However, there may be options that should be applied to the Compressor itself (for example, any options with effect beyond the Compress method). How to manage those options in a safe manner is something I still need to close on. It could be part of PR #6717.

mrambacher reviewed (commented):
Generally I like where this is heading.  I have some questions on CompressionInfo (whether it is turned upside down enough) and on the CompressorRegistry but I think this is a very good thing to have.

lucagiac81 commented on include/rocksdb/compressor.h:
```diff
@@ -0,0 +1,206 @@
+
+#include "memory/memory_allocator.h"
+namespace ROCKSDB_NAMESPACE {
+class Compressor;
```
Forward declaration moved.

lucagiac81 commented on include/rocksdb/compressor.h:
```diff
@@ -0,0 +1,206 @@
+  std::string dict_;
+  void* processed_dict_ =
+      nullptr;  // if the compressor supports processing the dictionary
+  std::shared_ptr<Compressor> compressor_ = nullptr;
```
Changed CompressionDict and UncompressionDict from struct to class. I think this is cleaner for the public API, and it is consistent with CompressionContext and UncompressionContext.
Added comments for members.

lucagiac81 commented on include/rocksdb/compressor.h:
```diff
@@ -0,0 +1,206 @@
+  std::shared_ptr<Compressor> compressor_ = nullptr;
+
+ public:
+  CompressionDict(std::string dict, CompressionType type, int level);
```
Fixed

lucagiac81 commented on include/rocksdb/compressor.h:
```diff
@@ -0,0 +1,206 @@
+
+  std::shared_ptr<Compressor> compressor_ = nullptr;
+
+  UncompressionDict(std::string dict, CompressionType type);
```
Fixed

lucagiac81 commented on include/rocksdb/compressor.h:
```diff
@@ -0,0 +1,206 @@
+  std::shared_ptr<Compressor> compressor_ = nullptr;
+
+  UncompressionDict(std::string dict, CompressionType type);
+  UncompressionDict(Slice slice, CacheAllocationPtr&& allocation,
```
Fixed

lucagiac81 commented on include/rocksdb/compressor.h:
```diff
@@ -0,0 +1,206 @@
+  CompressionType type() const { return type_; }
+};
+
+class Compressor {
+ public:
```
Added comments following doxygen/javadoc format (same used in Configurable and Customizable classes, for example).

lucagiac81 commented on include/rocksdb/compressor.h:
```diff
@@ -0,0 +1,206 @@
+
+  virtual const char* Name() const = 0;
+
+  virtual bool Supported() const { return true; }
```
The Supported method may be used by a compressor to verify any dependencies or configuration needed. I think most (custom) compressors would not need to override this method. If a compressor made it to the registry, all necessary checks may have already been done (e.g., while loading a shared library). 
Built-in compressors are different: they are always added to the registry, and dependencies are checked through ifdef guards in the Supported method.
Making it a pure virtual method is not a problem, if that's preferable.
A similar comment applies to DictCompressionSupported (added after PR #7619).

lucagiac81 commented:
Addressed some of the comments above and rebased branch. There are two additional changes from the previous version.
- compressor.h: added DictCompressionSupported method in Compressor (needed after PR #7619).
- options_helper.cc: GetSupportedCompressions and GetSupportedDictCompressions iterate over CompressorRegistry types, rather than using compression_type_string_map.

lucagiac81 commented on include/rocksdb/memory_allocator.h:
```diff
@@ -36,6 +36,21 @@ class MemoryAllocator {
   }
 };
 
+inline char* Allocate(size_t size, MemoryAllocator* allocator) {
```
Added comments

lucagiac81 commented on util/compression.cc:
```diff
@@ -0,0 +1,273 @@
+                                 uint32_t compress_format_version,
+                                 const char* input, size_t input_length,
+                                 std::string* output) {
+  bool success = BZip2_Compress(*info, compress_format_version, input,
+                                input_length, output);
```
If Bzip2 is not found, the compressor is still added to the registry, but Supported() returns false. The Compress/Uncompress methods in BZip2Compressor call BZip2Compress/BZip2Uncompress (original functions in compression.h). Those functions are guarded by ifdef. I didn't move the code from those functions to reduce code changes.
The same comment applies to other compressors.

lucagiac81 commented on util/compression.cc:
```diff
@@ -0,0 +1,273 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+//
+#include "util/compression.h"
```
compression.h still contains the implementation of Compress/Uncompress for the different algorithms (see previous comment).

lucagiac81 commented on util/compression.cc:
```diff
@@ -0,0 +1,273 @@
+                                    MemoryAllocator* allocator) {
+  (void)info;
+  (void)compress_format_version;
+  (void)allocator;
```
Commented out unused arg names.

lucagiac81 commented on table/block_based/block_based_table_reader.cc:
```diff
@@ -89,7 +89,7 @@ class BlocklikeTraits<BlockContents> {
                                size_t /* read_amp_bytes_per_bit */,
                                Statistics* /* statistics */,
-                               bool /* using_zstd */,
+                               CompressionType /* compression */,
```
Once a compressor is in the registry, both its name and its numeric type uniquely identify it. As the numeric type is already used in the rest of the code, I opted for it.
When serializing data to files, the compressor name must be recorded, as the numeric type assigned by the registry may change across RocksDB instances.

lucagiac81 commented on table/block_based/block_based_table_reader.cc:
```diff
@@ -849,11 +850,9 @@ Status BlockBasedTable::ReadPropertiesBlock(
-               CompressionTypeToString(kZSTD) ||
-           rep_->table_properties->compression_name ==
-               CompressionTypeToString(kZSTDNotFinalCompression));
+      rep_->compression_type = static_cast<CompressionType>(
+          CompressorRegistry::NewInstance()->GetCompressorType(
+              rep_->table_properties->compression_name));
```
Refer to previous comment.

lucagiac81 commented on util/compression.h:
```diff
@@ -515,54 +219,27 @@ inline bool ZSTDNotFinal_Supported() {
     case kDisableCompressionOption:
       return "DisableOption";
     default:
+      std::shared_ptr<Compressor> compressor =
+          CompressorRegistry::NewInstance()->GetCompressor(compression_type);
+      if (compressor != nullptr) {
+        return compressor->Name();
+      }
```
I'd like to keep the Compressor class generic, with no reference to specific compressors. Once we add support for custom compressors, we won't know all the names in advance. 
Currently, each compressor is responsible for providing its own name. However, there isn't a defined mapping between a compressor name and its class name.

lucagiac81 commented on util/compression.h:
```diff
@@ -643,6 +320,22 @@ inline CacheAllocationPtr Snappy_Uncompress(
 #endif
 }
 
```
kNoCompression is treated as a special case in many places, and it was easier to leave it that way. Also, the NoCompressor would never be reached for Compress/Uncompress calls.
kDisableCompressionOption is used in CompactionPicker to indicate that level-specific compression types should be used. As it's also treated as a special case, I left it unchanged.

lucagiac81 replied:
NoCompressor was added in the latest code

pdillinger commented:
Keep in mind that for us to accept this, the performance penalty would have to be minimal.

lucagiac81 commented:
Absolutely. Performance is a key requirement for this feature. From the data I've collected so far with db_bench (readrandom and readrandomwriterandom workloads), the performance difference is in general <1% (throughput, p99 latency). I'm collecting more data covering different conditions and additional metrics, so we can have a more accurate picture.
Are there specific conditions you'd recommend for performance testing, or do you have a set of benchmarks you already use to validate the effect of any code changes?

mrambacher commented on db/blob/blob_file_reader_test.cc:
```diff
@@ -706,9 +711,10 @@ TEST_F(BlobFileReaderTest, Compression) {
-                              &bytes_read));
+    ASSERT_OK(
+        reader->GetBlob(read_options, key, blob_offset, blob_size,
+                        BuiltinCompressor::GetCompressor(kSnappyCompression),
```
Get it once and store it?  Maybe at the top and replace the Snappy_Supported with compressor->Supported?

lucagiac81 replied:
Updated per suggestion (reuse compressor and use Supported method)

mrambacher commented on db/column_family.cc:
```diff
@@ -117,24 +117,28 @@ void GetIntTblPropCollectorFactory(
+                                   " is not linked with the binary.");
+  } else if (!moptions.compressor_per_level.empty()) {
+    for (const auto& compressor : moptions.compressor_per_level) {
+      if (!compressor->Supported()) {
```
Is there any possibility that this compressor is null?

lucagiac81 replied:
- compressor_per_level is initialized as an empty vector. 
- MutableCFOptions::RefreshDerivedOptions populates compressor_per_level using the content of compression_per_level.
- Compressors are obtained from BuiltinCompressor::GetCompressor, which could return nullptr in case of an invalid selection in compression_per_level.

I thought an invalid selection in compression_per_level would be caught before reaching this method, but that's incorrect. I added a null check and a test in compression_test (to verify the expected error is returned in case of invalid selection in compression_per_level).

mrambacher commented on db/db_block_cache_test.cc:
```diff
@@ -1991,10 +1982,20 @@ TEST_P(DBBlockCachePinningTest, TwoLevelDB) {
-  // current behavior of persisting/accessing dictionary even when there's no
-  // (de)compression happening, which seems fairly likely to change over time.
-  options.compression = kNoCompression;
+  // Select one of the available libraries that support dictionary compression
+  if (Zlib_Supported()) {
+    options.compression = kZlibCompression;
+  } else if (LZ4_Supported()) {
```
Is there a reason this does not use the first of Compressor::GetDictSupported() ?

lucagiac81 replied:
Updated to use Compressor::GetDictSupported() and pick the first available library that supports dictionary compression. The same was done in table_test (IncreaseCacheReservationForCompressDictBuildingBufferOnBuilderAddAndDecreaseOnBuilderFinish)

mrambacher commented on db/db_test.cc:
```diff
@@ -5010,7 +5010,8 @@ TEST_F(DBTest, DynamicLevelCompressionPerLevel2) {
         Compaction* compaction = reinterpret_cast<Compaction*>(arg);
         if (compaction->output_level() == 4) {
-          ASSERT_TRUE(compaction->output_compression() == kLZ4Compression);
+          ASSERT_EQ(compaction->output_compression()->GetCompressionType(),
```
1.  I wonder if output_compression should be renamed to compressor or output_compressor?
2. Does it make sense to have a compaction->GetCompressionType shortcut method?

lucagiac81 replied:
Renamed output_compression to output_compressor.
I did not add the shortcut method for compression type. I see both GetCompressionType and GetId called on the compressor returned by output_compressor in different places.

mrambacher commented on db/db_test2.cc:
```diff
@@ -1529,10 +1534,11 @@ TEST_P(PresetCompressionDictTest, Flush) {
       ASSERT_LE(TestGetTickerCount(options,
                                    BLOCK_CACHE_COMPRESSION_DICT_BYTES_INSERT),
-                2 * kBlockLen);
+                2 * kBlockLen + sizeof(UncompressionDict));
```
Is this meant to be the sizeof the Uncompression Dictionary itself or the struct?

lucagiac81 replied:
In a previous version of the code, I added a Compressor* member to UncompressionDict, and these tests started failing due to the size being slightly larger than 2 blocks. With the current code, these changes are actually not needed, and they were removed.

mrambacher commented on db/flush_job.cc:
```diff
@@ -82,22 +82,25 @@ const char* GetFlushReasonString (FlushReason flush_reason) {
+                   SnapshotChecker* snapshot_checker, JobContext* job_context,
+                   LogBuffer* log_buffer, FSDirectory* db_directory,
+                   FSDirectory* output_file_directory,
+                   std::shared_ptr<Compressor> output_compression,
```
const & ?

lucagiac81 replied:
Updated. Also renamed output_compression to output_compressor (to match the naming used in Compaction).

lucagiac81 commented on include/rocksdb/compressor.h:
```diff
@@ -0,0 +1,302 @@
+  // @param output Compressed data.
+  // Returns OK if compression completed correctly.
+  // Returns other status in case of error (e.g., Corruption).
+  virtual Status Compress(const CompressionInfo* info,
```
Changed to reference

lucagiac81 commented on include/rocksdb/compressor.h:
```diff
@@ -0,0 +1,302 @@
+  // Compress data.
+  // @param info Pointer to CompressionInfo object (containing context,
+  // dictionary, options).
+  // @param compress_format_version Compressed block format.
```
Format version is now part of CompressionInfo

lucagiac81 commented on include/rocksdb/compressor.h:
```diff
@@ -0,0 +1,302 @@
+  // @param output Compressed data.
+  // Returns OK if compression completed correctly.
+  // Returns other status in case of error (e.g., Corruption).
+  virtual Status Compress(const CompressionInfo* info,
```
This is addressed in the latest code. CompressionInfo doesn't include a Compressor or CompressionContext. In particular, if a compressor needs to carry context (such as ZSTDCompressor), it is responsible for managing it internally rather than through a generic CompressionContext class.

lucagiac81 commented on include/rocksdb/compressor.h:
```diff
@@ -0,0 +1,302 @@
+  // Returns OK if uncompression completed correctly.
+  // Returns other status in case of error (e.g., Corruption).
+  virtual Status Uncompress(const UncompressionInfo* info,
+                            uint32_t compress_format_version, const char* input,
```
Similar solution as for Compress. MemoryAllocator is now in UncompressionInfo.

lucagiac81 commented on include/rocksdb/compressor.h:
```diff
@@ -0,0 +1,302 @@
+  // @param level Compression level.
+  // Returns a pointer to the processed dictionary.
+  virtual void* ProcessCompressionDict(std::string& dict, int level) {
+    (void)dict;
```
This method was replaced by NewCompressionDict and dict is const &

lucagiac81 commented on include/rocksdb/compressor.h:
```diff
@@ -0,0 +1,302 @@
+  // Similar to ProcessCompressionDict, but for the uncompression dictionary.
+  // @see Compressor:ProcessCompressionDict for more details.
+  virtual void* ProcessUncompressionDict(Slice& dict) {
+    (void)dict;
```
This method was replaced by NewUncompressionDict and reference made const

lucagiac81 commented on include/rocksdb/compressor.h:
```diff
@@ -0,0 +1,302 @@
+
+  // Similar to ProcessCompressionDict, but for the uncompression dictionary.
+  // @see Compressor:ProcessCompressionDict for more details.
+  virtual void* ProcessUncompressionDict(Slice& dict) {
```
The updated NewCompressionDict and NewUncompressionDict methods take the same parameters (const string&)

lucagiac81 commented on include/rocksdb/compressor.h:
```diff
@@ -0,0 +1,302 @@
+  // @param processed_dict Pointer to processed dictionary.
+  virtual void DestroyCompressionDict(void* processed_dict) {
+    (void)processed_dict;
+  }
```
This method is removed. There is a ProcessedDict class that a Compressor can subclass to hold a processed/digested dictionary. The destructor of that class takes care of releasing any resources.

lucagiac81 commented on include/rocksdb/compressor.h:
```diff
@@ -0,0 +1,302 @@
+
+  // Get the memory usage for the processed uncompression dictionary.
+  // @param processed_dict Pointer to processed dictionary.
+  virtual size_t GetUncompressionDictMemoryUsage(void* processed_dict) {
```
This method is removed from Compressor

lucagiac81 commented on db/blob/blob_file_reader.cc:
```diff
@@ -47,11 +47,12 @@ Status BlobFileReader::Create(
-  CompressionType compression_type = kNoCompression;
+  std::shared_ptr<Compressor> compressor =
+      BuiltinCompressor::GetCompressor(kNoCompression);
 
```
Initialization is not needed. It was present in the original code with CompressionType and it was preserved. It's now removed.

lucagiac81 commented on db/blob/blob_file_reader.cc:
```diff
@@ -47,11 +47,12 @@ Status BlobFileReader::Create(
+      BuiltinCompressor::GetCompressor(kNoCompression);
 
   {
     const Status s = ReadHeader(file_reader.get(), column_family_id, statistics,
```
The Status part was outside the scope of the PR, and it was left unchanged.

lucagiac81 commented on db/blob/blob_file_reader.cc:
```diff
@@ -179,7 +180,7 @@ Status BlobFileReader::ReadHeader(const RandomAccessFileReader* file_reader,
   }
 
-  *compression_type = header.compression;
+  *compressor = BuiltinCompressor::GetCompressor(header.compression);
```
The assignment is for the shared_ptr, not the managed object (and GetCompressor already returns a shared_ptr). I think it looks ok.

lucagiac81 commented on db/blob/blob_file_reader.cc:
```diff
@@ -270,11 +271,11 @@ Status BlobFileReader::ReadFromFile(const RandomAccessFileReader* file_reader,
 BlobFileReader::BlobFileReader(
     std::unique_ptr<RandomAccessFileReader>&& file_reader, uint64_t file_size,
-    CompressionType compression_type, SystemClock* clock,
+    std::shared_ptr<Compressor> compressor, SystemClock* clock,
```
Changed to const &

lucagiac81 commented on db/blob/blob_file_reader.cc:
```diff
@@ -285,7 +286,7 @@ BlobFileReader::~BlobFileReader() = default;
                                const Slice& user_key, uint64_t offset,
                                uint64_t value_size,
-                               CompressionType compression_type,
+                               std::shared_ptr<Compressor> compressor,
```
Changed to const std::shared_ptr&

lucagiac81 commented on db/blob/blob_file_reader.cc:
```diff
@@ -534,33 +535,31 @@ Status BlobFileReader::VerifyBlob(const Slice& record_slice,
   assert(value);
 
-  if (compression_type == kNoCompression) {
+  if (compressor->GetCompressionType() == kNoCompression) {
```
Added

lucagiac81 commented on db/blob/blob_file_reader.cc:
```diff
@@ -534,33 +535,31 @@ Status BlobFileReader::VerifyBlob(const Slice& record_slice,
   constexpr uint32_t compression_format_version = 2;
   constexpr MemoryAllocator* allocator = nullptr;
+  UncompressionInfo info(UncompressionDict::GetEmptyDict(),
+                         compression_format_version, allocator);
```
Good point. Arguments match the default constructor. Updated.

lucagiac81 commented on db/blob/blob_file_reader.h:
```diff
@@ -54,13 +55,13 @@ class BlobFileReader {
       autovector<PinnableSlice*>& values, uint64_t* bytes_read) const;
 
-  CompressionType GetCompressionType() const { return compression_type_; }
+  std::shared_ptr<Compressor> GetCompressor() const { return compressor_; }
```
Updated to const std::shared_ptr &

lucagiac81 commented on db/blob/blob_file_reader.h:
```diff
@@ -42,7 +43,7 @@ class BlobFileReader {
   Status GetBlob(const ReadOptions& read_options, const Slice& user_key,
                  uint64_t offset, uint64_t value_size,
-                 CompressionType compression_type,
+                 std::shared_ptr<Compressor> compressor,
```
Updated

lucagiac81 commented on db/blob/blob_file_reader.h:
```diff
@@ -73,7 +74,7 @@ class BlobFileReader {
   static Status ReadHeader(const RandomAccessFileReader* file_reader,
                            uint32_t column_family_id, Statistics* statistics,
-                           CompressionType* compression_type);
+                           std::shared_ptr<Compressor>* compressor);
```
Updated

lucagiac81 commented on db/blob/blob_file_reader.h:
```diff
@@ -54,13 +55,13 @@ class BlobFileReader {
  private:
   BlobFileReader(std::unique_ptr<RandomAccessFileReader>&& file_reader,
-                 uint64_t file_size, CompressionType compression_type,
+                 uint64_t file_size, std::shared_ptr<Compressor> compressor,
```
Updated

lucagiac81 commented on db/blob/blob_file_reader_test.cc:
```diff
@@ -186,8 +183,9 @@ TEST_F(BlobFileReaderTest, CreateReaderAndGetBlob) {
-                              blob_sizes[0], kNoCompression, prefetch_buffer,
-                              &value, &bytes_read));
+                              blob_sizes[0],
+                              BuiltinCompressor::GetCompressor(kNoCompression),
```
It makes sense. It's used several times in CreateReaderAndGetBlob. Updated.

lucagiac81 commented:
I ran fillseq and readrandom benchmarks, comparing the PR against the previous commit on the main branch. 
Workload parameters: 16-byte keys, 256-byte values, 4kB blocks, ZSTD/LZ4 compression, db on /dev/shm, no block cache.

The relative ops/s differences (PR vs main, mean of 5 runs) are listed below, although they are not statistically significant:
- fillseq: +0.07% (ZSTD), -0.1% (LZ4)
- readrandom: -0.18% (ZSTD), +0.51% (LZ4)

The instruction path length is also verified to be matched within 0.2%. 
As we make changes as part of the review, we can verify performance stays matched with the latest code.

mrambacher commented on test_util/testutil.cc:
```diff
@@ -191,7 +193,7 @@ std::string RandomName(Random* rnd, const size_t len) {
 
 CompressionType RandomCompressionType(Random* rnd) {
   auto ret = static_cast<CompressionType>(rnd->Uniform(6));
```
Not that this is your bug, but 6 seems wrong here.  The builtin types go from 0x0 to 0x7 (so 8 of them)

lucagiac81 commented on util/compressor.cc:
```diff
@@ -0,0 +1,378 @@
+Status Compressor::CreateDictionary(
+    std::vector<std::string>& data_block_buffers,
+    std::unique_ptr<CompressionDict>* compression_dict) {
+  uint32_t max_dict_bytes = GetMaxDictBytes();
```
assert added for compression_dict (in BlockBasedTableBuilder::EnterUnbuffered, where CreateDictionary is called). Added check for DictCompressionSupported in CreateDictionary. Also added comment for IsDictionaryEnabled to return false if DictCompressionSupported is false.

lucagiac81 commented on util/compressor.cc:
```diff
@@ -0,0 +1,378 @@
+                                     compression_dict_sample_lens,
+                                     max_dict_bytes, level);
+    }
+    compression_dict->reset(NewCompressionDict(dict));
```
PR #6717 already introduced a separate SampleAndTrainDictionary method to make it easier for plugins to override only parts of CreateDictionary (and use the default sampling/training methods inside a customized implementation of CreateDictionary). I now split sampling and training into two virtual methods (SampleDict and TrainDict). TrainDict's default implementation still uses ZSTD_TrainDictionary and ZSTD_FinalizeDictionary, so plugins are able to use those as default.
I also renamed CreateDictionary as CreateDict and IsDictionaryEnabled as IsDictEnabled for consistency with other methods.
Some more tweaks may be needed to finalize the interface.

lucagiac81 commented on util/compressor.cc:
```diff
@@ -0,0 +1,378 @@
+#endif  // ROCKSDB_LITE
+  return max_dict_bytes;
+}
+
```
I agree. All compressors so far override these methods. The methods are now changed to simply return a default value. This also eliminates the different behavior for the LITE version, which doesn't support GetOption.

lucagiac81 commented on util/compressor.h:
```diff
@@ -0,0 +1,446 @@
+
+  // Returns a new compression dictionary from the input dict.
+  // Classes which have a ProcessedDict should override this method.
+  virtual CompressionDict* NewCompressionDict(const std::string& dict) {
```
Simple change. CreateDictionary was already placing the raw pointer in a std::unique_ptr. We cannot do the same for UncompressionDict without further changes, as the Create method of BlocklikeTraits<UncompressionDict>returns a raw pointer and that's aligned with other specializations.

lucagiac81 commented on tools/sst_dump_tool.cc:
```diff
@@ -247,12 +247,24 @@ int SSTDumpTool::Run(int argc, char const* const* argv, Options options) {
+        } else {
+          ConfigOptions config_options;
+          std::shared_ptr<Compressor> compressor;
+          Compressor::CreateFromString(config_options, compression_type,
+                                       &compressor);
+          if (compressor == nullptr) {
```
Updated

lucagiac81 commented on tools/db_bench_tool.cc:
```diff
@@ -2810,12 +2808,13 @@ class Benchmark {
+          BuiltinCompressor::GetCompressor(FLAGS_compression_type_e);
+      CompressionInfo info(CompressionDict::GetEmptyDict(),
+                           compress_format_version,
                            FLAGS_sample_for_compression);
```
Modified CompressionDict constructor to take sample_for_compression as parameter with default value. This made the code in db_bench_tool.cc more concise.

lucagiac81 commented on tools/db_bench_tool.cc:
```diff
@@ -3945,23 +3944,23 @@ class Benchmark {
+        BuiltinCompressor::GetCompressor(FLAGS_compression_type_e);
+    auto raw_compressor = compressor.get();
+    CompressionInfo info(CompressionDict::GetEmptyDict(),
+                         compress_format_version, FLAGS_sample_for_compression);
```
Same comment as above

lucagiac81 commented on tools/db_bench_tool.cc:
```diff
@@ -3977,25 +3976,22 @@ class Benchmark {
+                                     compress_format_version,
+                                     FLAGS_sample_for_compression);
+    UncompressionInfo uncompression_info(UncompressionDict::GetEmptyDict(),
+                                         compress_format_version);
```
Same comment as above

lucagiac81 commented on utilities/blob_db/blob_db_impl.cc:
```diff
@@ -1162,12 +1164,13 @@ Status BlobDBImpl::DecompressSlice(const Slice& compressed_value,
+    auto compressor = BuiltinCompressor::GetCompressor(compression_type);
+    UncompressionInfo info(
+        UncompressionDict::GetEmptyDict(),
+        GetCompressFormatForVersion(kBlockBasedTableVersionFormat));
```
Yes it is. Updated

japui-coder reviewed (commented):
utilities/cache_dump_load_impl.cc

japui-coder reviewed (commented):
utilities/cache_dump_load_impl.cc @
fuutilities/blob_db/blob_db_impl.cc. Patch .

japui-coder reviewed (commented):
db/db_block_cache_test.cc   . Patch .

pdillinger commented on table/block_based/block_based_table_reader.cc:
```diff
@@ -922,9 +916,17 @@ Status BlockBasedTable::ReadPropertiesBlock(
-          rep_->table_properties->compression_name !=
-          CompressionTypeToString(kNoCompression);
+      ConfigOptions config_options;
+      s = Compressor::CreateFromString(config_options,
```
I don't like that this implementation requires dependency injection (ObjectRegistry) (also known as "spooky action at a distance") to use a custom compressor. Reliable dependency injection can be a pain point for some people integrating RocksDB, e.g. through a shared object file.

Using prefix_extractor as an example, when the SST file uses the same prefix extractor as what is currently configured (for new SST files), dependency injection is not required to use it for reads.

lucagiac81 replied:
Compressor::CreateFromString will reuse an existing instance if an equivalent one is already available. The existing instances are tracked by the Compressor class rather than passed to the table like prefix_extractor. 

To follow a similar approach as for prefix_extractor, it should be possible to pass a custom compressor instance to RocksDB via options, without adding a factory function to ObjectRegistry. Would this address the concern with dependency injection? I will update PR #6717 (which adds support for custom compressors) to include and test this scenario.

pdillinger commented on table/block_based/block_based_table_reader.h:
```diff
@@ -725,6 +732,14 @@ struct BlockBasedTable::Rep {
   }
+
+  std::shared_ptr<Compressor> GetCompressor(CompressionType compression_type) {
+    if (compression_type == compressor->GetCompressionType()) {
```
Important to note that only one custom compression type can be used per file. Allowing various custom compressors in a single file would be a substantial schema+code change. Yes someone could add their own layer within the custom compressor, but that would be extra overhead.

We don't currently foresee needing various compressions within a file (except for some blocks *not* compressed) but we want to be sure everyone is on board with this decision.

lucagiac81 replied:
Yes, that is a limitation for plugin compressors. For built-in compressors, it's still possible to have different compressors per file. Their CompressionType id is still stored in each block as before. Plugins, on the other hand, all have the same id and rely on the compressor name in the properties block.

If various custom compressors per file are needed, we may be able to handle it by assigning "dynamic" numeric ids to custom compressors and recording them in blocks as usual. I can look into the implementation details if there is interest.

pdillinger replied:
> we may be able to handle it by assigning "dynamic" numeric ids to custom compressors and recording them in blocks as usual

Right. Might be worth keeping this in mind as a possible future feature, especially for schema purposes. I think the ideal state right now would be to reserve what that would look like in the schema, detect on read if that future feature is in use and provide a good error message.

pdillinger commented on table/block_based/block_based_table_reader_sync_and_async.h:
```diff
@@ -285,11 +285,15 @@ DEFINE_SYNC_AND_ASYNC(void, BlockBasedTable::RetrieveMultipleBlocks)
-        s = UncompressSerializedBlock(
-            info, req.result.data() + req_offset, handle.size(), &contents,
-            footer.format_version(), rep_->ioptions, memory_allocator);
+        std::shared_ptr<Compressor> compressor =
```
Creating temporary copies of shared_ptrs on the read "hot path" is a performance hazard.

In the past we have had to re-engineer some new customizable integrations because of excessive shared_ptr ops.

lucagiac81 replied:
The copy of shared_ptr<Compressor> is needed here in case the block is compressed by a different compressor than the one specified for the table (in rep_). I'll try to remove the copy at least for the most common case (block compressed with the table compressor).

pdillinger replied:
Yeah, I think that's an over-simplified interface. There should be immortal (de-)compressors for built-in algorithms. And custom compressors should be kept alive by the table reader (shared_ptr). (Are the compressors thread-safe by the way?)

While it might be nice to have a "GetCompressor" that copies a shared_ptr for arbitrary lifetime extension, the one used most would be to get a (de-)compressor reference or pointer that we promise not to use beyond the life of the table reader (which is guaranteed by the table cache to survive pending read operations).

pdillinger commented on table/block_based/block_based_table_reader.cc:
```diff
@@ -152,15 +152,16 @@ Status ReadAndParseBlockFromFile(
     const PersistentCacheOptions& cache_options,
-    MemoryAllocator* memory_allocator, bool for_compaction, bool async_read) {
+    MemoryAllocator* memory_allocator, bool for_compaction,
+    const std::shared_ptr<Compressor>& compressor, bool async_read) {
```
I consider `const shared_ptr<X>&` parameters to be an anti-pattern. It prevents move semantics where it might be useful, and is non-committal about intention to save a reference. I suggest deciding between `const X&` and plain `shared_ptr<X>`.

lucagiac81 replied:
I see the "mixed message" of passing const shared_ptr<T>&. I'll review all occurrences and verify if they're justified (e.g., if a copy of the shared_ptr is eventually made).

pdillinger commented on util/compressor.h:
```diff
@@ -0,0 +1,493 @@
+  UncompressionDict& operator=(const CompressionDict&) = delete;
+};
+
+// Interface for each compression algorithm to implement.
```
What are the thread safety expectations?

pdillinger commented:
By the way, we are finally very interested in this line of features. Thanks for working on it!

lucagiac81 commented:
> By the way, we are finally very interested in this line of features. Thanks for working on it!

That's great, thank you! I'll work on the items above and let's continue the review. I'll also rebase, as I see some new conflicts.

lucagiac81 commented:
Hi. We are planning to resume the pluggable compression work covered by this PR and PR #6717. We'd like to confirm there is still interest in driving this forward. If so, we will rebase on the latest version, address the feedback above, and coordinate to review when ready. Please let us know. Looking forward to completing this feature and making it available in RocksDB!
